clang/lib/Sema/SemaType.cpp

```diff
@@ -7438,6 +7438,60 @@ static bool HandleWebAssemblyFuncrefAttr(TypeProcessingState &State,
   return false;
 }
 
+static void HandleSwiftAttr(TypeProcessingState &State, TypeAttrLocation TAL,
+                            QualType &QT, ParsedAttr &PAttr) {
+  if (TAL == TAL_DeclName)
+    return;
+
+  Sema &S = State.getSema();
+  auto &D = State.getDeclarator();
+
+  // If the attribute appears in declaration specifiers
+  // it should be handled as a declaration attribute,
+  // unless it's associated with a type or a function
+  // prototype (i.e. appears on a parameter or result type).
+  if (State.isProcessingDeclSpec()) {
```

Comment:
This is a change in the current interpretation, right?  Isn't `swift_attr` currently allowed in the decl-specifiers and always applied to the declaration, even on function declarations?

Reply:
Yes, I expected that switching from `InheritedAttr` to `DeclOrTypeAttr` would preserve the behavior and attributes that appear on the outer decl-spec of a parameter  would be handled by SemaDeclAttr but that is not the case. I couldn't figure out a robust way to determine the position because there appears to be no way to walk up declarators so I opted for a simpler approach where anything that appears in "prototype" context is handled as a type attribute and adjusted that on the Swift side too.

Reply:
Okay, Pavel explained to me offline that the Swift importer already has to walk into the type-as-written to look for `AttributedType`s that might have been written at the top level of a declarator (where they syntactically apply to the outermost declarator; this is common with e.g. function declarators, where people often write attributes after the prototype).  In that case, making the rule *always* prefer to apply attributes to the type, and then making the importer also look for "declaration" attributes in nested positions seems reasonable.  (As ever, the real problem here is that the C grammar is poorly designed for interpreting this sort of thing with universal rules.)